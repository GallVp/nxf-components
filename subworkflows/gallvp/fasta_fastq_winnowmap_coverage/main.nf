include { MERYL_COUNT                   } from '../../../modules/gallvp/meryl/count/main'
include { WINNOWMAP                     } from '../../../modules/gallvp/winnowmap/main'
include { PAFTOOLS_SAM2PAF              } from '../../../modules/gallvp/paftools/sam2paf/main'
include { T2TPOLISH_PAFTOCOVCLIPPEDWIG  } from '../../../modules/gallvp/t2tpolish/paftocovclippedwig/main'

workflow FASTA_FASTQ_WINNOWMAP_COVERAGE {

    take:
    ch_fasta                            // channel: [ val(meta), fasta ]
    ch_fastq                            // channel: [ val(meta2), fasta ]
                                        // Combine by: meta2.ref_id = meta.id
    val_k                               // Integer; Default: 15
    val_meryl_distinct                  // Float; Default: 0.9998
    val_coverage_span                   // Integer; Default: 1024

    main:

    ch_versions                         = Channel.empty()

    // MODULE: MERYL_COUNT
    MERYL_COUNT ( ch_fastq, val_k ?: 15 )
    ch_versions                         = ch_versions.mix(MERYL_COUNT.out.versions.first())

    // MODULE: WINNOWMAP
    ch_win_inputs                       = ch_fastq
                                        | join ( MERYL_COUNT.out.meryl_db )
                                        | map { meta2, fastq, meryl_db -> [ meta2.ref_id, fastq, meryl_db ] }
                                        | combine(
                                            ch_fasta.map { meta, fasta -> [ meta.id, meta, fasta ] },
                                            by: 0
                                        )
                                        | multiMap { _ref_id, fastq, meryl_db, meta, fasta ->
                                            reads: [ meta, fastq ]
                                            reference: [ meta, fasta ]
                                            meryl_db: [ meta, meryl_db ]
                                        }
    WINNOWMAP (
        ch_win_inputs.reads,
        ch_win_inputs.reference,
        ch_win_inputs.meryl_db,
        val_meryl_distinct ?: 0.9998,
        true, // bam_format
        false, // bam_index_extension
        true, // cigar_paf_format
        true, // cigar_paf_format
        false, // sort_bam
    )

    ch_versions                         = ch_versions.mix(WINNOWMAP.out.versions.first())

    // MODULE: PAFTOOLS_SAM2PAF
    PAFTOOLS_SAM2PAF (
        WINNOWMAP.out.bam
    )

    ch_versions                         = ch_versions.mix(PAFTOOLS_SAM2PAF.out.versions.first())

    // MODULE: T2TPOLISH_PAFTOCOVCLIPPEDWIG
    ch_t2t_inputs                       = PAFTOOLS_SAM2PAF.out.paf
                                        | multiMap { meta, paf ->
                                            paf: [ meta, paf ]
                                            name: "${meta.id}"
                                            span: val_coverage_span ?: 1024
                                        }
    T2TPOLISH_PAFTOCOVCLIPPEDWIG (
        ch_t2t_inputs.paf,
        ch_t2t_inputs.name,
        ch_t2t_inputs.span
    )

    ch_versions                         = ch_versions.mix(T2TPOLISH_PAFTOCOVCLIPPEDWIG.out.versions.first())


    emit:
    wig                                 = T2TPOLISH_PAFTOCOVCLIPPEDWIG.out.cov  // channel: [ val(meta), wig ]
    versions                            = ch_versions                           // channel: [ versions.yml ]
}
