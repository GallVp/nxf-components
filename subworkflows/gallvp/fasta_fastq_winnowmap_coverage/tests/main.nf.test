nextflow_workflow {

    name "Test Subworkflow FASTA_FASTQ_WINNOWMAP_COVERAGE"
    script "../main.nf"
    config "./nextflow.config"
    workflow "FASTA_FASTQ_WINNOWMAP_COVERAGE"

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/fasta_fastq_winnowmap_coverage"

    tag "meryl/count"
    tag "paftools/sam2paf"
    tag "winnowmap"
    tag "t2tpolish/paftocovclippedwig"

    test("sarscov2 - fasta - fastq") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'test', single_end:true, ref_id: 'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ])
                input[2] = 15
                input[3] = 0.9998
                input[4] = 1024
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("sarscov2 - fasta - fastq - stub") {

        options '-stub'

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'test', single_end:true, ref_id: 'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ])
                input[2] = 15
                input[3] = 0.9998
                input[4] = 1024
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
