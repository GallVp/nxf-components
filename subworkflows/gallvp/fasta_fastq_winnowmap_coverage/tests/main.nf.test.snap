{
    "sarscov2 - fasta - fastq - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_ref"
                        },
                        "test_ref.cov.wig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,00a207ae678bdd38c77f8bf4a5fe5bfe",
                    "versions.yml:md5,a8ed9cb8b5731afa1a1f100ee0f3244b",
                    "versions.yml:md5,eced9bce07e4ba513d5ad7eeb03c0288",
                    "versions.yml:md5,f82212ab895ddbe7ba7a55cb37b77022"
                ],
                "versions": [
                    "versions.yml:md5,00a207ae678bdd38c77f8bf4a5fe5bfe",
                    "versions.yml:md5,a8ed9cb8b5731afa1a1f100ee0f3244b",
                    "versions.yml:md5,eced9bce07e4ba513d5ad7eeb03c0288",
                    "versions.yml:md5,f82212ab895ddbe7ba7a55cb37b77022"
                ],
                "wig": [
                    [
                        {
                            "id": "test_ref"
                        },
                        "test_ref.cov.wig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-30T13:36:56.884879"
    },
    "sarscov2 - fasta - fastq": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_ref"
                        },
                        "test_ref.cov.wig:md5,720c8e2b21a9cf4a5aebb5a44da3a590"
                    ]
                ],
                "1": [
                    "versions.yml:md5,00a207ae678bdd38c77f8bf4a5fe5bfe",
                    "versions.yml:md5,58c7e3ce8c7f41ed641a00261997c444",
                    "versions.yml:md5,7d7b976fc0769125ad74ea4b3cc11ee0",
                    "versions.yml:md5,eced9bce07e4ba513d5ad7eeb03c0288"
                ],
                "versions": [
                    "versions.yml:md5,00a207ae678bdd38c77f8bf4a5fe5bfe",
                    "versions.yml:md5,58c7e3ce8c7f41ed641a00261997c444",
                    "versions.yml:md5,7d7b976fc0769125ad74ea4b3cc11ee0",
                    "versions.yml:md5,eced9bce07e4ba513d5ad7eeb03c0288"
                ],
                "wig": [
                    [
                        {
                            "id": "test_ref"
                        },
                        "test_ref.cov.wig:md5,720c8e2b21a9cf4a5aebb5a44da3a590"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-30T13:35:14.993452"
    }
}